internals-js/src/definitions.ts

```diff
@@ -3197,20 +3197,40 @@ export class Directive<
   }
 }
 
-export function sameDirectiveApplication(application1: Directive<any, any>, application2: Directive<any, any>): boolean {
-  return application1.name === application2.name && argumentsEquals(application1.arguments(), application2.arguments());
+/**
+ * Checks if 2 directive applications should be considered equal.
+ *
+ * By default, 2 directive applications are considered equal if they are for the same directive and are passed the same values to
+ * the same arguments. However, some special directive can be excluded so that no 2 applications are ever consider equal. By default,
+ * this is the case of @defer, as never want to merge @defer applications so that each create its own "deferred block".
+ */
+export function sameDirectiveApplication(
+  application1: Directive<any, any>,
+  application2: Directive<any, any>,
+  directivesNeverEqualToThemselves: string[] = [ 'defer' ],
+): boolean {
+  // Note: we check name equality first because this method is most often called with directive that are simply not the same
+  // name and this ensure we exit cheaply more often than not.
+  return application1.name === application2.name 
```

Comment:
I think this code gets called both during operations and also during composition. I'm assuming that it's safe to use the `application.name` because in order to compose, directives would need to be named the same in the supergraph? Just want to make sure this won't provide false positives if directives have been renamed.

Reply:
> I'm assuming that it's safe to use the application.name because in order to compose, directives would need to be named the same in the supergraph?

Correct. And if we do ever allow merging directives that have different names in different subgraphs, we will definitively have to change how we handle this. And fwiw, I'll note that this part of the logic is not changed by this patch: all that I'm truly changing here is that we skip saying that 2 `@defer` application are ever "the same", and we can't have `@defer` applications during composition anyway (since it's not allowed on any type system locations), so this patch should be safe on that front.